pkg/kubelet/config/common.go

```diff
@@ -43,25 +43,47 @@ import (
 	"github.com/golang/glog"
 )
 
+type staticPod struct {
+	nodeName types.NodeName
+	pod      *api.Pod
+}
+
 // Generate a pod name that is unique among nodes by appending the nodeName.
 func generatePodName(name string, nodeName types.NodeName) string {
 	return fmt.Sprintf("%s-%s", name, strings.ToLower(string(nodeName)))
 }
 
 func applyDefaults(pod *api.Pod, source string, isFile bool, nodeName types.NodeName) error {
+	if pod.Annotations == nil {
+		pod.Annotations = make(map[string]string)
+	}
+
 	if len(pod.UID) == 0 {
+		staticpod := &staticPod{
+			nodeName: nodeName,
+			pod:      pod,
+		}
+		// Generate pod UID.
 		hasher := md5.New()
+		hash.DeepHashObject(hasher, staticpod)
+		pod.UID = types.UID(hex.EncodeToString(hasher.Sum(nil)[0:]))
+		glog.V(5).Infof("Generated UID %q pod %q from %s", pod.UID, pod.Name, source)
+
+		// Generate legacy hash.
```

Comment:
Please explain why this is needed and when it's safe to remove the code.

Reply:
Seems like this doesn't actually help to fix the issue as per your https://github.com/kubernetes/kubernetes/pull/57135#pullrequestreview-99907101. And it's never safe to remove the code as long as there are "legacy" pods and we don't want them to restart when upgrading kubelet...

/hold
until we have a final solution.